File: src/main/java/com/dyonovan/neotech/common/container/machine/ContainerCrafter.java
package com.dyonovan.neotech.common.container.machine;

import com.dyonovan.neotech.common.container.BaseContainer;
import com.dyonovan.neotech.common.tileentity.machine.TileEntityCrafter;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.*;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.CraftingManager;

public class ContainerCrafter extends BaseContainer {
    protected TileEntityCrafter crafter;

    public InventoryCrafting craftingGrid1;
    public InventoryCrafting craftingGrid2;

    public IInventory craftResult1 = new InventoryCraftResult();
    public IInventory craftResult2 = new InventoryCraftResult();

    public ContainerCrafter(InventoryPlayer playerInv, TileEntityCrafter tile) {
        this.crafter = tile;
        craftingGrid1 = new DummyCraftingInventory(tile, this);
        craftingGrid2 = new DummyCraftingInventory(tile, this, 10);

        this.addSlotToContainer(new SlotCrafting(playerInv.player, craftingGrid1, craftResult1, 0, 80, 31));
        this.addSlotToContainer(new SlotCrafting(playerInv.player, craftingGrid2, craftResult2, 1, 80, 59));

        addCraftingGrid(craftingGrid1, 0,  8, 27, 3, 3);
        addCraftingGrid(craftingGrid2, 0, 116, 27, 3, 3);

        bindPlayerInventory(playerInv, 8, 84);

        onCraftMatrixChanged(craftingGrid1);
        onCraftMatrixChanged(craftingGrid2);
    }

    public void onCraftMatrixChanged(IInventory inv) {
        if(inv == craftingGrid1)
            craftResult1.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(craftingGrid1, crafter.getWorld()));
        if(inv == craftingGrid2)
            craftResult2.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(craftingGrid2, crafter.getWorld()));
    }

    @Override
    public boolean canMergeSlot(ItemStack stack, Slot slot) {
        return !(slot instanceof SlotCrafting);
    }


    public void addCraftingGrid(IInventory inventory, int startSlot, int x, int y, int width, int height) {
        int i = 0;
        for(int h = 0; h < height; h++) {
            for(int w = 0; w < width; w++) {
                this.addSlotToContainer(new Slot(inventory, startSlot + i++, x + (w * 18), y + (h * 18)));
            }
        }
    }

    @Override
    public boolean canInteractWith(EntityPlayer playerIn) {
        return true;
    }

    public static class DummyCraftingInventory extends InventoryCrafting {

        protected final TileEntityCrafter tile;
        protected final int offset;
        protected final Container container;
        protected ItemStack[] stacks;

        public DummyCraftingInventory(TileEntityCrafter tile, Container container) {
            this(tile, container, 0);
        }

        public DummyCraftingInventory(TileEntityCrafter tile, Container container, int offset) {
            super(null, 3, 3);
            stacks = new ItemStack[9];
            this.tile = tile;
            this.offset = offset;
            this.container = container;
        }

        private void onCraftingChanged() {
            container.onCraftMatrixChanged(this);
        }

        @Override
        public int getSizeInventory() {
            return 9;
        }

        @Override
        public ItemStack getStackInSlot(int slot) {
            return slot >= getSizeInventory() ? null : tile.getStackInSlot(slot + offset);
        }

        @Override
        public ItemStack getStackInRowAndColumn(int row, int column) {
            if (row >= 0 && row < 3) {
                int k = row + column * 3;
                return getStackInSlot(k);
            } else
                return null;
        }

        @Override
        public ItemStack getStackInSlotOnClosing(int slot) {
            return tile.getStackInSlotOnClosing(slot + offset);
        }

        @Override
        public ItemStack decrStackSize(int slot, int size) {
            ItemStack stack = tile.decrStackSize(slot + offset, size);
            onCraftingChanged();
            return stack;
        }

        @Override
        public void setInventorySlotContents(int slot, ItemStack stack) {
            tile.setInventorySlotContents(slot + offset, stack);
            onCraftingChanged();
        }
    }
}

File: src/main/scala/com/teambrmodding/neotech/common/container/misc/ContainerCrafter.scala
package com.teambrmodding.neotech.common.container.misc

import com.teambrmodding.neotech.common.tiles.misc.TileCrafter
import com.teambr.bookshelf.common.container.BaseContainer
import net.minecraft.entity.player.{EntityPlayer, InventoryPlayer}
import net.minecraft.inventory._
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.CraftingManager

import scala.collection.mutable.ArrayBuffer

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 18, 2015
 */
class ContainerCrafter(playerInventory: InventoryPlayer, tile: TileCrafter) extends BaseContainer(playerInventory, tile) {

    var craftResult1 = new InventoryCraftResult
    var craftResult2 = new InventoryCraftResult

    var crafter = tile
    var craftingGrid1: InventoryCrafting = new DummyCraftingInventory(tile, this)
    var craftingGrid2: InventoryCrafting = new DummyCraftingInventory(tile, this, 10)

    addSlotToContainer(new SlotCrafting(playerInventory.player, craftingGrid1, craftResult1, 0, 80, 31))
    addSlotToContainer(new SlotCrafting(playerInventory.player, craftingGrid2, craftResult2, 1, 80, 59))

    addCraftingGrid(craftingGrid1, 0, 8, 27, 3, 3)
    addCraftingGrid(craftingGrid2, 0, 116, 27, 3, 3)
    addPlayerInventorySlots(8, 84)

    onCraftMatrixChanged(craftingGrid1)
    onCraftMatrixChanged(craftingGrid2)

    override def onCraftMatrixChanged(inv: IInventory): Unit = {
        if(inv == craftingGrid1)
            craftResult1.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(craftingGrid1, crafter.getWorld))
        if(inv == craftingGrid2)
            craftResult2.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(craftingGrid2, crafter.getWorld))
    }

    override def canMergeSlot(stack: ItemStack, slot: Slot): Boolean = !slot.isInstanceOf[SlotCrafting]

    def addCraftingGrid(inventory: IInventory, startSlot: Int, x: Int, y: Int, width: Int, height: Int): Unit = {
        var i: Int = 0
        for (h <- 0 until height) {
            for(w <- 0 until width) {
                addSlotToContainer(new Slot(inventory, startSlot + i, x + (w * 18), y + (h * 18)))
                i += 1
            }
        }
    }

    override def canInteractWith(player: EntityPlayer): Boolean = true

    override def transferStackInSlot(player: EntityPlayer, slotId: Int): ItemStack = {
        println(slotId)
        val slot: Slot = inventorySlots.get(slotId).asInstanceOf[Slot]
        if (slot != null && slot.getHasStack) {
            val itemToTransfer: ItemStack = slot.getStack
            val copy: ItemStack = itemToTransfer.copy
            if (slotId < inventorySize) {
                if (!mergeItemStackSafe(itemToTransfer, inventorySize, inventorySlots.size, reverse = true)) return null
            }
            else if (!mergeItemStackSafe(itemToTransfer, 0, inventorySize, reverse = false)) return null
            if (itemToTransfer.stackSize == 0) slot.putStack(null)
            else slot.onSlotChanged()
            if (itemToTransfer.stackSize != copy.stackSize) {
                slot.onPickupFromSlot(player, copy)
                return copy
            }
        }
        null
    }

    class DummyCraftingInventory(tile: TileCrafter, container: Container, offset: Int)
            extends InventoryCrafting(null, 3, 3) {

        def this(tile: TileCrafter, container: Container) = this(tile, container, 0)

        val stacks = new ArrayBuffer[ItemStack]

        private def onCraftingChanged(): Unit = container.onCraftMatrixChanged(this)

        override def getSizeInventory: Int = 9

        override def getStackInSlot(slot: Int): ItemStack = {
            if (slot >= getSizeInventory) null else tile.getStackInSlot(slot + offset)
        }

        override def getStackInRowAndColumn(row: Int, column: Int): ItemStack = {
            if (row >= 0 && row < 3) {
                val k: Int  = row + column * 3
                getStackInSlot(k)
            } else
                null
        }

        //override def getStackInSlotOnClosing(slot: Int): ItemStack = tile.getStackInSlotOnClosing(slot + offset)

        override def decrStackSize(slot: Int, size: Int): ItemStack = {
            val stack:ItemStack = tile.decrStackSize(slot + offset, size)
            onCraftingChanged()
            stack
        }

        override def setInventorySlotContents(slot: Int, stack: ItemStack): Unit = {
            tile.setInventorySlotContents(slot + offset, stack)
            onCraftingChanged()
        }
    }
}
